controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java

```diff
@@ -37,16 +72,207 @@ public String getName() {
 
     @Override
     public CompletableFuture<Void> deleteScope() {
-        return store.deleteNode(scopePath);
+        return store.deleteNode(scopePath)
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(counterPath), DATA_NOT_FOUND_PREDICATE, null))
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(streamsInScopePath), DATA_NOT_FOUND_PREDICATE, null));
+    }
+
+    CompletableFuture<Void> addStreamToScope(String name, int streamPosition) {
```

Comment:
I asked below whether we can do this with hashing the stream name, in which case, we don't need the sequential znode.

But, I'm also concerned about this use of a sequential znode. Which new sequential znode you create is a new child to the same parent, so this is limiting us to have at most 1 million streams. It is pretty large number, but it might be unnecessarily capping the number of streams we can have.

Reply:
> But, I'm also concerned about this use of a sequential znode. Which new sequential znode you create is a new child to the same parent, so this is limiting us to have at most 1 million streams. It is pretty large number, but it might be unnecessarily capping the number of streams we can have.

We are capping the number of streams under a scope to a 10 digit number. Which is essentially Integer.Max ==> 2 billion.

Reply:
PS: my original implementation was based on a hashing scheme itself.. but that had several nuances. This hierarchy approach is much cleaner and has monotonically increasing stream position so continuation token is fairly straightforward. only caveat is to store the stream position for deletion.

Reply:
I don't think this is right. In `getNextStreamPosition`, we are creating a sequential znode under the same parent. Each time we do that, it will create a new znode with an incremented value appended to the name. There is a limit on the number of children of a parent znode, which is 1 million. I think your 2 billion reference is talking about the ids, while I'm talking about the number of children under a parent znode.

Reply:
The counter is an ephemeral sequential znode. 
But yes, theoretically we can have more than a million concurrent ephemeral znodes under the same scope. So i could delete the counter ephemeral node immediately after i create it.